packages/router/src/grpc/custody/authorize/index.ts

```diff
@@ -1,19 +1,25 @@
-import type { Impl } from '.';
-import { extLocalCtx, extSessionCtx } from '../../ctx/prax';
-import { approverCtx } from '../../ctx/approver';
+import type { Impl } from '..';
+import { extLocalCtx, extSessionCtx, servicesCtx } from '../../../ctx/prax';
+import { approverCtx } from '../../../ctx/approver';
 import { generateSpendKey } from '@penumbra-zone/wasm/src/keys';
 import { authorizePlan } from '@penumbra-zone/wasm/src/build';
 import { Key } from '@penumbra-zone/crypto-web/src/encryption';
 import { Code, ConnectError } from '@connectrpc/connect';
 import { Box } from '@penumbra-zone/types/src/box';
 import { UserChoice } from '@penumbra-zone/types/src/user-choice';
+import { assertValidSwaps } from './assert-valid-swaps';
+import { isControlledAddress } from '@penumbra-zone/wasm/src/address';
 
 export const authorize: Impl['authorize'] = async (req, ctx) => {
   if (!req.plan) throw new ConnectError('No plan included in request', Code.InvalidArgument);
 
   const approveReq = ctx.values.get(approverCtx);
   const sess = ctx.values.get(extSessionCtx);
   const local = ctx.values.get(extLocalCtx);
+  const services = ctx.values.get(servicesCtx);
+
+  const { fullViewingKey } = (await services.getWalletServices()).viewServer;
```

Comment:
I personally find await's nested in things hard to read. What do you think about?

```typescript
  const services = await ctx.values.get(servicesCtx).getWalletServices();

  assertValidSwaps(req.plan, address =>
    isControlledAddress(services.viewServer.fullViewingKey, address),
  );
```

Reply:
```suggestion
  const { viewServer: { fullViewingKey } } = await services.getWalletServices();
```

but more generally, we should probably just have a context item for fvk

Reply:
Good call, done